src/bot.py

```diff
@@ -1,71 +1,76 @@
-            # Positive radians in the unit circle is a turn to the left.
-            turn = -1.0  # Negative value for a turn to the left.
-            action_display = "turn left"
-        else:
-            turn = 1.0
-            action_display = "turn right"
+        """
+        This function will be called by the framework many times per second. This is where you can
+        see the motion of the ball, etc. and return controls to drive your car.
+        """
 
-        self.controller_state.throttle = 1.0
-        self.controller_state.steer = turn
+        # Start the renderer. Make sure you call end_rendering at the end of this function.
+        self.renderer.begin_rendering()
 
-        draw_debug(self.renderer, my_car, packet.game_ball, action_display)
+        # This is good to keep at the beginning of get_output. It will allow you to continue
+        # any sequences that you may have started during a previous call to get_output.
+        if self.active_sequence and not self.active_sequence.done:
+            return self.active_sequence.tick(packet)
 
-        return self.controller_state
-
-
-def find_correction(current: Vec3, ideal: Vec3) -> float:
-    # Finds the angle from current to ideal vector in the xy-plane. Angle will be between -pi and +pi.
-
-    # The in-game axes are left handed, so use -x
-    current_in_radians = math.atan2(current.y, -current.x)
-    ideal_in_radians = math.atan2(ideal.y, -ideal.x)
-
-    diff = ideal_in_radians - current_in_radians
+        # Gather some information about our car and the ball
+        my_car = packet.game_cars[self.index]
+        car_location = Vec3(my_car.physics.location)
+        car_velocity = Vec3(my_car.physics.velocity)
+        ball_location = Vec3(packet.game_ball.physics.location)
 
-    # Make sure that diff is between -pi and +pi.
-    if abs(diff) > math.pi:
-        if diff < 0:
-            diff += 2 * math.pi
+        if car_location.dist(ball_location) > 1000:
+            # We're far away from the ball, let's try to lead it a little bit
+            ball_prediction = self.get_ball_prediction_struct()  # This can predict bounces, etc
+            ball_in_future = find_slice_at_time(ball_prediction, packet.game_info.seconds_elapsed + 2)
+            target_location = Vec3(ball_in_future.physics.location)
         else:
-            diff -= 2 * math.pi
-
-    return diff
-
-
-def draw_debug(renderer, car, ball, action_display):
-    renderer.begin_rendering()
-    # draw a line from the car to the ball
-    renderer.draw_line_3d(car.physics.location, ball.physics.location, renderer.white())
-    # print the action that the bot is taking
-    renderer.draw_string_3d(car.physics.location, 2, 2, action_display, renderer.white())
-    renderer.end_rendering()
+            target_location = ball_location
+
+        # Draw some things to help understand what the bot is thinking
+        self.renderer.draw_line_3d(car_location, target_location, self.renderer.white())
```

Comment:
When leading the ball, we render a line from the car to a future ball position. This is correct, except it might look like a bug, because it seemingly connects our car with nothing.

Maybe rendering a red rectangle at the target or rending two seconds of the ball prediction might make it clearer?

Reply:
Good point! I'll probably do both if it doesn't make too much mess in the code

Reply:
Looks much better, thanks!